### PR TITLE
Get v1 data, partial implementation

### PR DESCRIPTION
**Pull request by e-kotov**
Hi @Robinlovelace , I have finished some of the work on v1 data and tried to envision how it would be united with current code for v2, as well as how I would suggest the v2 code may be changed and improved upon. Could you please review what I have for v1 (and minor changes to v2) below?

Setup:

```{r}
remotes::install_github("Robinlovelace/spanishoddata@get-v1-data")
Sys.setenv(SPANISH_OD_DATA_DIR = "/your/cache/path")
```

#### V1 Data

Fetch available data for version 1 with an optional check for local files:

```{r}
v1 <- spod_available_data_v1(check_local_files = TRUE)
```

This function changes the default paths of the MITMA data locally to arrange it in a hive style for faster filtering and/or import with `duckdb/arrow`.

#### Function Wrapping Suggestion

Consider wrapping both functions below into a single function with a version argument, for example, `spod_available_data(ver = 1)`. Alternatively, you could name them `spod_available_data_v1()` and `spod_available_data_v2()`, or `spod_get_metadata_v1()` and `spod_get_metadata_v2()`. Another option could be `spod_metadata_v1()` and `spod_metadata_v2()`. The current function handles the XML file cache by checking when it was last downloaded and fetches it automatically if it is older than one day. It might make sense to add an argument to load the latest cached XML for offline use.

#### V2 Data

Fetch metadata for version 2:

```{r}
v2 <- spod_get_metadata()
```

I have not modified this function much, but I plan to add the path-changing feature for v2 data as well if it is appreciated in v1. The few changes I have made include updating the XML file name so it differs from v1, renaming the file mask to `data_links`, and removing the `current_timestamp` argument from the XML fetching function, as the current timestamp should be used automatically.

The naming of both functions is still up for discussion, and I am open to suggestions.

#### Get Latest V1 File List

```{r}
v1_file <- spod_get_latest_v1_file_list()
```

This is only really for internal use or for advanced users, so I would have this explicit long name and change to similar for v2, or even combine the functions into one.

#### Available Data for V1

```{r}
v1 <- spod_available_data_v1() # with optional file check for local files as noted in the beginning of the message
```

#### Zones Loader

The zones loader with English parameter names and their variations that are converted to Spanish using internal helper functions. These functions load the shapefiles, check the geometry validity, and replace the uppercase ID with lowercase 'id'. In the future, I plan to attach metadata such as municipality names.

```{r}
zones_municip <- spod_get_zones_v1("muni")
zones_municip <- spod_get_zones_v1("municip")
zones_municip <- spod_get_zones_v1("municipalities")
zones_districts <- spod_get_zones_v1("dist")
zones_districts <- spod_get_zones_v1("distr")
zones_districts <- spod_get_zones_v1("districts")
```

### V2 Zones

For v2 zones, more adjustments are needed, such as attaching accompanying names and population numbers metadata that are available alongside the files. This is a TODO.

### Download Function

The new download function handles most non-spatial data, automatically checks for the required data version, accepts multiple types of zoning, and processes date arguments automatically.

```{r}
spod_download_data(type = "od", zones = "districts", dates = "2020032[0-5]")
spod_download_data(type = "od", zones = "municip", dates = "20200302_20200305")
spod_download_data(type = "od", zones = "municip", dates = "2020-03-02_2020-03-05")
spod_download_data(type = "od", zones = "municip", dates = c(start = "2020-03-02", end = "2020-03-05"))
```

To download any two dates without interpreting them as a range, use the following. More than two dates are always interpreted as an arbitrary sequence and not converted to a range.

```{r}
spod_download_data(type = "od", zones = "municip", dates = c("2020-03-02", "2020-03-05"))
```

I think this advanced handling of dates would make the package more accessible for beginners, as I do not expect everyone to easily grasp what a regex is and they may be more comfortable using simpler dates argument.

There are also "testthat" tests for the underlying function that handles the magic behind the dates argument. That includes tests for data incompatibility (i.e. the helper functions check if the requested date range is spanning across multiple data versions and stops with a warning).

### Next Steps

The next step is to implement an equivalent of `spod_get_zones()` and similar functions for v1 data that would rely on the already downloaded data or download it automatically if it is missing, this will use the same mechanics that I implemented for the download function as they would simply wrap it. I would appreciate your feedback on the current implementation and naming of the functions, as well as my suggestions on the overall structure of the package.

We can discuss during our next scheduled meeting.

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Good stuff @e-kotov, thanks for the clear statement of changes and options, will test in due course, likely on Friday.

**Comment by e-kotov** [comment]
@Robinlovelace great! Take your time. Meanwhile, I added some last minute additions to finally pass the R CMD check. We now have a snapshot of metadata bundled with the package (compressed xmls are tiny and `xml2::read_xml()` reads them directly from gz just fine). These can be used for future internal tests of some other functions too.

**Comment by Robinlovelace** [comment]
Sounds good, more soon!

**Comment by Robinlovelace** [comment]
OK, going to give this a quick test and hopefully review now. Have quite limited time, ~15 minutes, due to family commitments later in the day.

**Comment by Robinlovelace** [comment]
Update here: looking great!

File: DESCRIPTION [hunk 1]
```diff
@@ -29,10 +29,14 @@ Imports:
     lubridate,
     purrr,
     readr,
+    rlang (>= 1.1.0),
```
Comment (Robinlovelace) [inline DESCRIPTION:32]: :+1: good package with few deps

File: NAMESPACE [hunk 2]
```diff
@@ -1,6 +1,10 @@
 # Generated by roxygen2: do not edit by hand
 
+export(spod_available_data_v1)
+export(spod_download_data)
 export(spod_get)
+export(spod_get_latest_v1_file_list)
```
Comment (Robinlovelace) [inline NAMESPACE:6]: :+1:

File: R/download_data.R [hunk 3]
```diff
@@ -0,0 +1,106 @@
+#' Download the data files of specified type, zones, and dates
```
Comment (Robinlovelace) [inline R/download_data.R:1]: Good to have a general dl function

File: R/download_data.R [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+#' # Download the origin-destination on district level for select dates in 2020 and 2021
+#' spod_download_data(type = "od", zones = "dist", 
+#'                    dates_list = c("2020-03-20", "2020-03-24", "2021-03-20", "2021-03-24"))
+#' 
+#' # Download the origin-destination on municipality level using regex for a date range in March 2020 
+#' # (the regex will capture the dates 2020-03-20 to 2020-03-24)
+#' spod_download_data(type = "od", zones = "municip", 
+#'                    date_regex = "2020032[0-4]")
+#' }
+spod_download_data <- function(
```
Comment (Robinlovelace) [inline R/download_data.R:29]: We should run an auto formatter, e.g. styler or styler.equals. That's for another day.

File: R/folders.R [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# change subfolder name for raw data cache here to apply globally
+spod_subfolder_raw_data_cache <- function(ver = 1) {
+  rlang:::check_number_whole(ver)
```
Comment (Robinlovelace) [inline R/folders.R:3]: Nice check function!

File: R/get.R [hunk 6]
```diff
@@ -30,6 +30,7 @@ spod_get_latest_v2_xml = function(
 #' This function retrieves the data dictionary for the specified data directory.
 #'
 #' @param data_dir The directory where the data is stored. Defaults to the value returned by `spod_get_data_dir()`.
+#' @param quiet Whether to suppress messages. Defaults to `FALSE`.
```
Comment (Robinlovelace) [inline R/get.R:33]: :+1:

File: R/get.R [hunk 7]
```diff
@@ -39,14 +40,14 @@ spod_get_latest_v2_xml = function(
 #' names(metadata)
 #' head(metadata)
 #' }
-spod_get_metadata = function(data_dir = spod_get_data_dir()) {
-  xml_files_list = fs::dir_ls(data_dir, type = "file", regexp = "data_links_") |> sort()
+spod_get_metadata = function(data_dir = spod_get_data_dir(), quiet = FALSE) {
+    xml_files_list = fs::dir_ls(data_dir, type = "file", regexp = "data_links_v2") |> sort()
```
Comment (Robinlovelace) [inline R/get.R:44]: :+1:

File: R/get.R [hunk 8]
```diff
@@ -74,12 +75,24 @@ spod_get_metadata = function(data_dir = spod_get_data_dir()) {
...
+#' If the environment variable is not set, it returns the temporary directory.
+#' 
+#' @return The data directory.
+#' @keywords internal
+spod_get_data_dir = function(quiet = FALSE) {
   data_dir_env = Sys.getenv("SPANISH_OD_DATA_DIR")
-  if (data_dir_env == "") {
-    data_dir_env = tempdir()
+  if( data_dir_env == "" ) {
+    if (isFALSE(quiet)) warning("Warning: SPANISH_OD_DATA_DIR is not set. Using the temporary directory, which is not recommended, as the data will be deleted when the session ends.\n\n To set the data directory, use `Sys.setenv(SPANISH_OD_DATA_DIR = '/path/to/data')` or set SPANISH_OD_DATA_DIR permanently in the environment by editing the `.Renviron` file locally for current project with `usethis::edit_r_environ('project')` or `file.edit('.Renviron')` or globally for all projects with `usethis::edit_r_environ('user')` or `file.edit('~/.Renviron')`.")
```
Comment (Robinlovelace) [inline R/get.R:88]: :+1: good warning message, super clear!

File: R/get_v1_data.R [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  # replace 2 digit day with 1 digit day
+  files_table$local_path <- gsub("day=0([1-9])", "day=\\1", files_table$local_path)
+
+  # now check if any of local files exist
+  files_table$downloaded <- fs::file_exists(files_table$local_path)
+
+  return(files_table)
+}
+
+#' Retrieves the zones for v1 data
```
Comment (Robinlovelace) [inline R/get_v1_data.R:125]: Great to have zones v1, different from v2 I guess.

File: R/get_v1_data.R [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  zones_path <- fs::dir_ls(data_dir, glob = glue::glue("**{zones}/*.shp"), recurse = TRUE)
+
+  zones <- spod_clean_zones_v1(zones_path)
+  fs::dir_create(fs::path_dir(expected_gpkg_path), recurse = TRUE)
+  sf::st_write(zones, expected_gpkg_path, delete_dsn = TRUE, delete_layer = TRUE)
+
+  return(zones)
+}
+
+#' Fixes common issues in the zones data and cleans up variable names
```
Comment (Robinlovelace) [inline R/get_v1_data.R:190]: Yes! So good to have this and will give the data more value to users.

**Review by Robinlovelace — APPROVED** [review]
This is an epic set of changes Egor, many thanks! Big +1, all tested on README and no further comment in limited time I have today. Let's get it merged to keep ball rolling but will leave that to you.

File: R/folders.R (hunk 5 above)
Comment (e-kotov) [reply R/folders.R:3]: the use of ::: to its internal functions raises warnings in cmd check, will try to follow the advise of https://josiahparry.com/posts/2024-06-30-type-safety/ to import the source code for these checks

File: R/download_data.R (hunk 4 above)
Comment (e-kotov) [reply R/download_data.R:29]: at some point yes, btw, I am a proponent of using `<-` for assignment in R.

File: R/get_v1_data.R (hunk 9 above)
Comment (e-kotov) [reply R/get_v1_data.R:125]: Internally the code for v1 and v2 zones will be very different. We can have `spod_get_zones(ver = 1)`, `spod_get_zones(ver = 2)`, as well as `spod_get_zones_v1()` and `spod_get_zones_v2()`. I think wrapper functions won't hurt.

**Comment by e-kotov** [comment]
Ok, merging, but new big updates are coming up)

File: R/download_data.R (hunk 4 above)
Comment (Robinlovelace) [reply R/download_data.R:29]: No objections, I realise I'm in the minority here, most developers are also arrow users. #2

File: R/folders.R (hunk 5 above)
Comment (e-kotov) [reply R/folders.R:3]: #23

**Comment by e-kotov** [comment]
Partially fulfils #5

**Comment by Robinlovelace** [comment]
1 version down, 1 to go!

**Pull request merged**